plugins/scaffolder-backend/sample-templates/v1beta2-demo/template.yaml

```diff
@@ -68,12 +74,19 @@ spec:
         repoUrl: '{{ parameters.repoUrl }}'
 
     - id: register
+      if: '{{ not parameters.dryRun }}'
       name: Register
       action: catalog:register
       input:
         repoContentsUrl: '{{ steps.publish.output.repoContentsUrl }}'
         catalogInfoPath: '/catalog-info.yaml'
 
+    - name: Results
+      if: '{{ parameters.dryRun }}'
+      action: publish:log
+      input:
+        listWorkspace: true
+
   output:
-    remoteUrl: '{{ steps.publish.output.remoteUrl }}'
-    entityRef: '{{ steps.register.output.entityRef }}'
+    remoteUrl: '{{#unless parameters.dryRun}}{{ steps.publish.output.remoteUrl }}{{/unless}}'
```

Comment:
Would it be a crazy to run handlebars in non strict mode and ignore variables it cannot find instead of wrapping templates with `unless`?

Reply:
Or, probably better, remove the `unless` and wrap the templating in a `try/catch` to return empty and log to either the `scaffolder-backend`.

Reply:
We probably need to do some different validation for things that will never exist (syntax errors) and things that might not exist (run time dependencies of steps)

Reply:
Disabling strict mode was something I first thought about, but it didn't feel right. 

If we try/catch the strict errors and log them to the task log, we would still have some kind of validation. I can take a look. I'm actually not sure what happens if the `remoteUrl` is empty, I have to check whether it is skipped in that case. In the case of try/catch, I would just set it to undefined.

Reply:
It's a bit strange, but `handlebars.compile` doesn't fail on syntax error, they are throw once you execute the template. But if you do a `handlebars.parse` before it, you could differentiate between syntax and runtime errors. Otherwise you could also try to check the error message, but that's not nice.

Reply:
Yeah, but I guess there's two types of syntax error right?

1) Handlebars syntax - e.g `something {{sdlands}`

2) Template errors, where we point to `parameters.foo` and foo does not exist

The first I think we capture already, but the second one probably need to validate seperately, and just try to execute the template in a try catch.

The other part of the validation that I didn't think about which could throw a pretty strange error message is that if you use that output in another step and handlebars tries to template it, but we catch the error and return '' I think it would fail the JSON schema validation for the actions as it's required but an empty string?

Reply:
So we could run the output like this:

```typescript

      const output = JSON.parse(
        JSON.stringify(task.spec.output),
        (_key, value) => {
          if (typeof value === 'string') {
            try {
              return this.handlebars.compile(value, {
                noEscape: true,
                strict: true,
                data: false,
                preventIndent: true,
              })(templateCtx);
            } catch (e) {
              if (e instanceof TypeError) {
                // Log here, but has to be sync!

                return undefined;
              }
              throw e;
            }
          }
          return value;
        },
      );
```

It would fallback to undefined if fields are missing, still throw syntax errors and allows to log missing fields. One problem is, that `task.emitLog` is an async function, but I can't await it in the `JSON.parse` callback... I could do fire and forget and not await it.

We could introduce the same approach also in the other places where handlebar templates are used (condition, inputs), but I'm not sure if we need them there. For condition they could be handy, to skip a step if a field is missing. For inputs it could be handy too, if one has an optional field that he don't want to fill if a field is missing. 

One the other hand, the logic for handling JSON differs between all three implementations.
* Condition: tries to parse JSON, falls back to a string - Useful if one wants to return boolean values
* Inputs: tries to only parse as JSON if it starts with either `{` (object) or `[` (array), falls back to string
* Outputs: template result is not interpreted as JSON.

I wonder if a common pattern would be better here.

Reply:
I run into another issue that also plays into the strict mode discussion: https://github.com/backstage/backstage/pull/5914

Reply:
I wonder if it just makes sense to turn off strict mode for everything then, and then log any errors to the console when templating fails. Hopefully the JSON Schema will take care of validating that the actions have the right parameters and stop running the action if it doesn't match the schema.

Reply:
I did an update and disabled strict mode. Without strict mode templates can be a lot simpler.

@benjdlambert Playing around with the error cases gives not so nice results. For example, if the JSON schema requires a string, it is happy for `{{ parameter.missing }}` as the template always evaluates to a string, but in this case an empty string. One would have to extend the JSON schema to require string with a `minLength`. Another alternative could be treating empty string results from templates in a special way. If we threat them as `undefined/null` JSON schema would detect it as expected. If one still needs to respond with an empty string, he could use `""` as the results are parsed as JSON (would require an update for the outputs).

---

I noticed another thing. `output.remoteUrl` seems to be deprecated and is replaced by `output.links`. However in that case it is much more difficult to have conditional outputs. Right now, an empty `output.remoteUrl` is skipped, here it is more difficult:

```yaml
output:
  links:
    - title: Download
      icon: archive
      url: '{{ steps.my-step.output.downloadUrl }}'
```

We could resolve it by filtering out all links that have no URL (where the `undefined` conversation from above would be handy again...)

Reply:
I implemented the idea with the handling empty strings as `undefined` and skipping links with `undefined` urls